db/product.go

```diff
@@ -0,0 +1,313 @@
+		logger.WithField("err", err.Error()).Error("Error fetching Product Ids from database")
+		return
+	}
+
+	// idArr stores id's of all products
+	var idArr []int
+
+	for result.Next() {
+		var Id int
+		err = result.Scan(&Id)
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Couldn't Scan Product ids")
+			return
+		}
+		idArr = append(idArr, Id)
+	}
+
+	for i := 0; i < len(idArr); i++ {
+		var product Product
+		product, err = s.GetProductByID(ctx, int(idArr[i]))
+		if err != nil {
+			logger.WithField("err", err.Error()).Error("Error selecting Product from database by id " + string(idArr[i]))
+			return
+		}
+		products = append(products, product)
+	}
+
+	return
+}
+
+// CreateNewProduct
+func (s *pgStore) CreateNewProduct(ctx context.Context, p Product) (createdProduct Product, err error) {
+	// First, make sure Product isn't already in db, if Product is present, just return the it
+	err = s.db.Get(&createdProduct, getProductByNameQuery, p.Name)
+	if err == nil {
+		// If there's already a product, err wil be nil, so no new Product is populated.
+		err = fmt.Errorf("Product Already exists!")
+		return
+	}
+	tx, err := s.db.Beginx() // Use Beginx instead of MustBegin so process doesn't die if there is an error
+	if err != nil {
+		// FAIL : Could not begin database transaction
+		logger.WithField("err", err.Error()).Error("Error beginning product insert transaction in db.CreateNewProduct with Id: " + string(p.Id))
+		return
+	}
+
+	_, err = tx.NamedExec(insertProductQuery, p)
+	//  p.Name, p.Description, p.Price, p.Discount, p.Tax, p.Quantity, p.CategoryId, p.Brand, p.Color, p.Size
+
+	if err != nil {
+		// FAIL : Could not run insert Query
+		logger.WithField("err", err.Error()).Error("Error inserting product to database: " + p.Name)
+		return
+	}
+	err = tx.Commit()
+	if err != nil {
+		// FAIL : transaction commit failed.Will Automatically rollback
+		logger.WithField("err", err.Error()).Error("Error commiting transaction inserting product into database: " + string(p.Id))
+		return
+	}
+
+	//length of url
+	urls := len(p.URLs)
+	var number int
+	//new insert record get id number
+	result, err := s.db.Query(newInsertRecord)
+	for result.Next() {
+		err = result.Scan(&number)
+	}
+
+	for i := 0; i < urls; i++ {
+		//insert urls of given records
+		_, err = s.db.Exec(insertProductURLsQuery, number, p.URLs[i])
+		if err != nil {
+			// FAIL : Could not run insert Query
+			logger.WithField("err", err.Error()).Error("Error inserting urls to database: ")
+			return
+		}
+	}
+
+	// Re-select Product and return it
+	createdProduct, err = s.GetProductByID(ctx, number)
+	if err != nil {
+		logger.WithField("err", err.Error()).Error("Error selecting from database with id: " + string(p.Id))
+		return
+	}
+	return
+}
+
+func (s *pgStore) UpdateProductStockById(ctx context.Context, product Product, Id int) (updatedProduct Product, err error) {
+
+	var dbProduct Product
+	err = s.db.Get(&dbProduct, getProductByIDQuery, Id)
```

Comment:
do we need to fetch product before updating it
can u check if update query fails if record does not exists

Reply:
Done there is no need to check a 2-time record is present.
if records are not found then return error record not found.